plugins/BEdita/Core/src/Job/Service/ThumbnailService.php

```diff
@@ -0,0 +1,55 @@
+<?php
+/**
+ * BEdita, API-first content management framework
+ * Copyright 2018 ChannelWeb Srl, Chialab Srl
+ *
+ * This file is part of BEdita: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * See LICENSE.LGPL or <http://gnu.org/licenses/lgpl-3.0.html> for more details.
+ */
+
+namespace BEdita\Core\Job\Service;
+
+use BEdita\Core\Filesystem\Thumbnail;
+use BEdita\Core\Job\JobService;
+use Cake\Datasource\Exception\RecordNotFoundException;
+use Cake\Log\Log;
+use Cake\ORM\TableRegistry;
+use Cake\Utility\Hash;
+
+/**
+ * Service to handle asynchronous generation of thumbnails.
+ *
+ * @since 4.0.0
+ */
+class ThumbnailService implements JobService
+{
+
+    /**
+     * {@inheritDoc}
+     */
+    public function run(array $payload, array $options = [])
+    {
+        try {
+            /** @var \BEdita\Core\Model\Table\StreamsTable $table */
+            $table = TableRegistry::get('Streams');
+            $stream = $table->get(Hash::get($payload, 'uuid'));
+
+            $generator = Thumbnail::getGenerator(Hash::get($payload, 'generator', 'default'));
+            $generator->generate($stream, (array)Hash::get($payload, 'options'));
+
+            return true;
+        } catch (RecordNotFoundException $e) {
+            // Stream not found, mark job as complete anyway.
```

Comment:
should we `Log` something or is this a normal flow?

Reply:
This happens when you enqueue a thumbnail-ization job for a given stream, and then delete that stream _before_ the job is processed. Is quite unusual, but the job should not be retried, because it is guaranteed to fail again. We still may want to log something — if it happens too often there almost certainly is a problem — but I wouldn't treat it as an error… maybe something like this?

```php
Log::debug(sprintf('Thumbnail service could not find stream %s', $stream->uuid));
```

Reply:
Seems ok, I'd use `Log::warn`since it can be related to an error but also `Log::info` is acceptable here...